src/main/java/org/jenkinsci/plugins/workflow/steps/CoreStep.java

```diff
@@ -78,16 +78,24 @@
         }
 
         @Override protected Void run() throws Exception {
-            final StepContext context = getContext();
-            final Run<?,?> run = context.get(Run.class);
+            final StepContext ctx = this.getContext();
+            final Run<?,?> run = ctx.get(Run.class);
             assert run != null;
-            final TaskListener listener = context.get(TaskListener.class);
+            final TaskListener listener = ctx.get(TaskListener.class);
             assert listener != null;
-            final EnvVars env = context.get(EnvVars.class);
+            final EnvVars env = ctx.get(EnvVars.class);
             assert env != null;
-            final FilePath workspace = context.get(FilePath.class);
-            final Launcher launcher = context.get(Launcher.class);
-            if (this.delegate.requiresWorkspace()) {
+            final FilePath workspace = ctx.get(FilePath.class);
+            final Launcher launcher = ctx.get(Launcher.class);
+            boolean workspaceRequired = true;
+            // TODO: Replace with a direct call to SimpleBuildStep.requiresWorkspace() once the minimum core version for this plugin is 2.25x or newer.
+            try {
+                final Method requiresWorkspace = this.delegate.getClass().getMethod("requiresWorkspace");
```

Comment:
```suggestion
                final Method requiresWorkspace = SimpleBuildStep.class.getMethod("requiresWorkspace");
```

and similarly elsewhere

Reply:
Wouldn’t that always call the default implementation instead of an override in the actual step instance?
I could see this for the `Disposer` case because that has a non-virtual one, but not for the `requiresWorkspace` on step and wrapper.

Reply:
The `Method` is defined in a particular class: `SimpleBuildStep`. `invoke` will run the implementation. There is no way in Java to do otherwise, other than by generating bytecode that calls `invokespecial`.

Reply:
No, `SimpleBuildStep` is an interface with a default implementation for `requiresWorkspace`. If you use `getMethod` on `SimpleBuildStep.class`, that default implementation is what you get and invoke, This breaks the processing (and the without-workspace-requirement test), because on an older Jenkins it will always throw `NoSuchMethodException`, causing it to require a workspace, and on a new Jenkins it always returns `true`, with the same effect.

It is essential to call the `getMethod("requiresWorkspace")` on the actual step type captured in `this.delegate`, so that it will find the override and invoke that.

The only case where this is not as important is on the Disposer(), because in new Jenkins, where the method exists, it is final. However, that causes a problem for the reflection-based approach, because that prevents any unit test for a case of a wrapper+disposer not requiring a workspace, because the disposer invocation would always think it required a workspace (due to the method not existing in `Disposer` itself). Given that I don't think a subclass can define a new method with the same signature as a final method from a superclass, also using the specific disposer type for the lookup should be harmless, and allows the unit test to work (and allows implementers of wrappers not to require a workspace when used in pipelines, even on old Jenkins).

Reply:
Really? Because

```java
import java.lang.reflect.Method;
public class M {
    public static void main(String[] args) throws Exception {
        Method m = I.class.getMethod("m");
        m.invoke(new C1());
        m.invoke(new C2());
    }
    interface I {
        default void m() {
            System.out.println("from I");
        }
    }
    static class C1 implements I {}
    static class C2 implements I {
        @Override public void m() {
            System.out.println("from C2");
        }
    }
}
```

prints

```
from I
from C2
```

as I would expect. Am I missing something?

Reply:
Hmmm - it may just be the unit tests that get broken then (because they have "fake" overrides).

Reply:
Yes I think that was the disconnect: https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/123#discussion_r461755336